File: LeetCode/Python3/Problem 122.py
# LeetCode: Problem 122
# Jonathan Kosasih

#Say you have an array for which the ith element is the price of a given stock on day i.
#
#Design an algorithm to find the maximum profit. You may complete as many transactions as you like
#(i.e., buy one and sell one share of the stock multiple times).
#
#Note: You may not engage in multiple transactions at the same time (i.e., you must sell the stock before you buy again).
#------------------------------------------------------------------
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if(not prices):
            return 0
        profit = 0
        for i in range(1,len(prices)):
            if(prices[i] > prices[i-1]):
                profit += prices[i] - prices[i-1]
        return profit
        
    
sol = Solution()
example = [7,1,5,3,6,4]
print(str(sol.maxProfit(example)))
example = [7,6,4,3,1]
print(str(sol.maxProfit(example)))
example = []
print(str(sol.maxProfit(example)))
example = [2,8,1,2]
print(str(sol.maxProfit(example)))

File: LeetCode/C++/Problem 122.cpp
// LeetCode: Problem 122
// Jonathan Kosasih
/*
Say you have an array for which the ith element is the price of a given stock on day i.

Design an algorithm to find the maximum profit. You may complete as many transactions as you like (i.e., buy one and sell one share of the stock multiple times).

Note: You may not engage in multiple transactions at the same time (i.e., you must sell the stock before you buy again).
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;

class Solution {
public:
	int maxProfit(vector<int>& prices) {
		if (prices.empty()) {
			return 0;
		}
		int profit = 0;
		for (int i = 1; i < prices.size(); i++) {
			if (prices[i] > prices[i - 1]) {
				profit += prices[i] - prices[i - 1];
			}
		}
		return profit;
	}
};
int main()
{
    //std::cout << "Hello World!\n";
}